### PR TITLE
[NTOS:KE/x64][SDK] Improve trap handling with some asm macros

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

This adds some asm macros for trap handling, uses them in the right places. It also fixes some problems, that existed before. The macros will be reused in other places.

## Proposed changes
- Add ASSERT_TRAP_FRAME_INTS_ENABLED, ASSERT_TRAP_FRAME_IRQL_VALID and ASSERT_IRQL_PASSIVE
- Add HANDLE_USER_APCS, make it versatile enough to be able to pass a register for the current thread and the trap frame

## Tests
- https://reactos.org/testman/compare.php?ids=94343,94347,94351,94366

### CODE REVIEW

File: sdk/include/asm/trapamd64.inc [hunk 1]
```diff
@@ -289,3 +306,4 @@ MACRO(TRAP_ENTRY, Trap, Flags)
     ENDFUNC
 ENDM
 
+
```
Comment (HBelusca) [inline sdk/include/asm/trapamd64.inc:309]: extra endline?
```suggestion
```

File: sdk/include/asm/trapamd64.inc [hunk 2]
```diff
@@ -1,3 +1,54 @@
...
+    LOCAL IrqlIsPassive
+    mov rax, cr8
+    test rax, rax
+    jz IrqlIsPassive
+    int HEX(2C)
+IrqlIsPassive:
+#endif
+ENDM
+
+// Checks for user APCs and delvers them if necessary.
```
Comment (HBelusca) [inline sdk/include/asm/trapamd64.inc:40]:
```suggestion
// Checks for user APCs and delivers them if necessary.
```

File: ntoskrnl/ke/amd64/trap.S [hunk 3]
```diff
@@ -839,19 +839,11 @@ GLOBAL_LABEL KiSystemCall64Again
...
-    jnz IntsEnabled
-    int 3
-IntsEnabled:
-#endif
+    ASSERT_TRAP_FRAME_INTS_ENABLED rsp + MAX_SYSCALL_PARAM_SIZE
 
-    /* Check for pending user APC */
+    /* Check for pending user APCs */
     mov rcx, gs:qword ptr [PcCurrentThread]
-    cmp byte ptr [rcx + ThApcState + AsUserApcPending], 0
```
Comment (HBelusca) [inline ntoskrnl/ke/amd64/trap.S:851]: ok, `ThApcState + AsUserApcPending == KTHREAD_UserApcPending`

File: ntoskrnl/ke/amd64/trap.S [hunk 4]
```diff
@@ -1140,19 +1132,33 @@ PUBLIC KiSetTrapContext
...
 
+    /* Save rax to not clobber the return for the system call handler */
+    mov [rsp + ExP1Home], rax
+
     /* Raise IRQL to APC_LEVEL */
     mov rax, APC_LEVEL
     mov cr8, rax
 
     /* Get the current thread */
     mov rbp, gs:[PcCurrentThread]
```
Comment (HBelusca) [inline ntoskrnl/ke/amd64/trap.S:1157]: i see elsewhere you've changed to `gs:qword ptr [PcCurrentThread]`

File: sdk/include/asm/trapamd64.inc [hunk 5]
```diff
@@ -1,3 +1,54 @@
...
+// Clobbers all volatile registers except rax.
+MACRO(HANDLE_USER_APCS, ThreadReg, TrapFrame)
+    LOCAL NoUserApcPending
+
+    /* Check for pending user APC */
+    cmp byte ptr [ThreadReg + ThApcState + AsUserApcPending], 0
```
Comment (HBelusca) [inline sdk/include/asm/trapamd64.inc:46]: Want to use `ThApiState + AsUserApcPending`, or `KTHREAD_UserApcPending` ? (even though both evaluate to the same thing)

**Review by HBelusca — APPROVED** [review]

File: sdk/include/asm/trapamd64.inc (hunk 1 above)
Comment (tkreuzer) [reply sdk/include/asm/trapamd64.inc:309]: Yes! Deal with it.

Comment (tkreuzer) [reply sdk/include/asm/trapamd64.inc]: OK, OK, I fixed it.

File: ntoskrnl/ke/amd64/trap.S (hunk 4 above)
Comment (tkreuzer) [reply ntoskrnl/ke/amd64/trap.S:1157]: It's not actually needed. Changed it to `gs:[PcCurrentThread]` now.

**Pull request merged**